client/my-sites/stats/activity-log-tasklist/index.jsx

```diff
@@ -0,0 +1,290 @@
+
+/**
+ * Checks if the supplied plugin is currently updating.
+ *
+ * @param {object} plugin Plugin to check status
+ * @returns {bool} True if plugin is updating.
+ */
+const isPluginUpdating = plugin => 'inProgress' === plugin.updateStatus.status;
+
+class ActivityLogTasklist extends Component {
+	static propTypes = {
+		siteId: PropTypes.number,
+		pluginsWithUpdates: PropTypes.arrayOf( PropTypes.object ), // Plugins with pending updates
+		allPluginsToUpdate: PropTypes.arrayOf( PropTypes.object ), // Don't use. Use connected pluginsToUpdate instead.
+
+		// Connected props
+		siteName: PropTypes.string.isRequired,
+		// Plugins already updated + those with pending updates.
+		// This extends allPluginsToUpdate with the plugin update status.
+		pluginsToUpdate: PropTypes.object.isRequired,
+
+		// Localize
+		translate: PropTypes.func.isRequired,
+		updateSinglePlugin: PropTypes.func.isRequired,
+		showErrorNotice: PropTypes.func.isRequired,
+		showInfoNotice: PropTypes.func.isRequired,
+		showSuccessNotice: PropTypes.func.isRequired,
+		removeThisNotice: PropTypes.func.isRequired,
+	};
+
+	state = {
+		pluginUpdateNotice: null, // keyed by plugin id, like "hello-dolly/hello"
+		dismissedPlugins: [],
+		dismissAll: false,
+	};
+
+	dismissAll = () => {
+		// ToDo: this should update some record in the tasklist API
+		this.setState( { dismissAll: true } );
+	};
+
+	dismissPlugin = pluginSlug => {
+		// ToDo: this should update some record in the tasklist API
+		this.setState( {
+			dismissedPlugins: this.state.dismissedPlugins.concat( [ pluginSlug ] ),
+		} );
+	};
+
+	updatePlugin = ( singlePlugin = false ) => {
+		const { removeThisNotice, showInfoNotice, siteName, updateSinglePlugin } = this.props;
+
+		// Use id: "hello-dolly/hello", slug: "hello-dolly", name: "Hello Dolly", updateStatus: bool|object,
+		const updateNoticeId = get(
+			this.state.pluginUpdateNotice,
+			[ singlePlugin.slug, 'notice', 'noticeId' ],
+			null
+		);
+
+		if ( updateNoticeId ) {
+			removeThisNotice( updateNoticeId );
+		}
+
+		updateSinglePlugin( singlePlugin );
+
+		this.setState( {
+			pluginUpdateNotice: merge( this.state.pluginUpdateNotice, {
+				[ singlePlugin.slug ]: showInfoNotice(
+					PluginNotices.inProgressMessage( 'UPDATE_PLUGIN', '1 site 1 plugin', {
+						plugin: singlePlugin.name,
+						site: siteName,
+					} ),
+					{
+						showDismiss: false,
+					}
+				),
+			} ),
+		} );
+	};
+
+	componentWillReceiveProps( nextProps ) {
+		if ( isEmpty( nextProps.pluginsToUpdate ) ) {
+			return;
+		}
+
+		const {
+			removeThisNotice,
+			showErrorNotice,
+			showSuccessNotice,
+			siteName,
+			translate,
+			pluginsToUpdate,
+		} = nextProps;
+		const newState = {};
+		const noticesToShow = {};
+		const pluginsCompleted = [];
+
+		each( Object.values( pluginsToUpdate ), plugin => {
+			const pluginSlug = plugin.slug;
+
+			if ( false === get( this.props.pluginsToUpdate, `${ pluginSlug }.updateStatus`, false ) ) {
```

Comment:
this is a general note on the use of `get()` and it applies to the other cases, but it's quite awkward here to be mixing around property access mechanisms and I'm not sure if I'm missing something or if there is a reason for it, but if we are going to use `get()` then I'd recommend staying all-in on one pattern (some exceptions like `this.props` or `this.state` are normal)…

```js
get( this.props, [ 'pluginsToUpdate', pluginSlug, 'updateStatus' ] )
```

also note that `get()` returns `undefined` if the value doesn't exist, so it _appears_ that the default value of `false` may be unnecessary

Reply:
Will test the value. The `updateStatus` is `false` before a update begins. When an update is under progress, it's an object.

I'll update the string in `get`. Can you point a reason not to compose a string like I did? Is it a guideline in Calypso? Perhaps this could be enforced in the pre-commit processor.

Reply:
It's not something likely easy to lint and there's no single reason not to do so - it's more of a general programming principle. In a single expression we have different coding patterns and it's not even adding any value. If it _were_ adding significant value we could say it's fine as an exception but in this case I'd argue that it's just making things needlessly inconsistent.

`get()` is designed to make deeply-nested property access safe, so if we are passing deeply-nested properties to its first argument we're taking away from its sole responsibility and jumping out of that safety. In this specific case we're fine because `this.props` will always exist and when `pluginsToUpdate` _doesn't_ exist it's still okay because `get()` knows how to handle `undefined` as the object.

`get()` is also designed to work with one of two modes: an array of path-traversals; and a dot-delimited string which it will parse and convert into an array of path-traversals. by using templates literals to build that dot-delimited path we're preemptively undoing what `get()` will do again. in other words, we take the end-result that `get()` wants, disguise it, then make `get()` uncover it to get back to where we started.

if we mainly want a statically-known deeply-nested property it's easy to pass a string

```js
get( data, 'items.camping.tents.0' )
```

if we are dynamically generating that path then the array format is more appropriate

```js
get( data, [ 'items', category, subCategory, 0 ] )
```